stranding/stranding.py

```diff
@@ -0,0 +1,190 @@
+import logging
+import warnings
+
+from Bio.pairwise2 import align, format_alignment
+from Bio.Seq import Seq
+
+from seqseek import Chromosome
+
+from .exceptions import (MissingReferenceFlank,
+                        InconsistentAlignment,
+                        Unstrandable,
+                        FlanksTooShort)
+
+
+LOGGER = logging.getLogger(__name__)
+DEFAULT_MIN_FLANK_LENGTH = 15
+DEFAULT_WINDOW_EXTENSION = 0
+
+
+FORWARD_STRAND = 1
+REVERSE_STRAND = -1
+
+
+# empirically derived default values from stranding hundreds of thousands of flanks
+# from an Illumina beadchip. Two points are awarded for each matching base and one
+# point is subtracted for each mismatch. Gaps are strongly discouraged with a 5 point
+# penalty.
+DEFAULT_MATCH_SCORE = 2
+DEFAULT_MISMATCH_PENALTY = -1
+DEFAULT_GAP_OPEN_PENALTY = -5
+DEFAULT_TOLERANCE = 0.77
+
+
+class GenomeStranding(object):
+
+    def __init__(self,
+                 min_flank_length=DEFAULT_MIN_FLANK_LENGTH,
+                 tolerance=DEFAULT_TOLERANCE,
+                 match_score=DEFAULT_MATCH_SCORE,
+                 mismatch_penalty=DEFAULT_MISMATCH_PENALTY,
+                 gap_open_penalty=DEFAULT_GAP_OPEN_PENALTY):
+
+            self.min_flank_length = min_flank_length
+            self.tolerance = tolerance
+            self.match_score = match_score
+            self.mismatch_penalty = mismatch_penalty
+            self.gap_open_penalty = gap_open_penalty
+            if self.min_flank_length < DEFAULT_MIN_FLANK_LENGTH:
+                warnings.warn('Short flank lengths may lead to inaccurate alignments')
+
+    def is_high_scoring(self, score, query):
+        if len(query) < self.min_flank_length:
+            return False
+        return score > len(query) * self.match_score * self.tolerance
+
+    def is_perfect_score(self, score, query):
+        if len(query) < self.min_flank_length:
+            return False
+        return score == len(query) * self.match_score
+
+    def align(self, ref, query, score_only=True):
+        alignment = align.localms(ref, query, self.match_score, self.mismatch_penalty,
+                                  self.gap_open_penalty, self.mismatch_penalty,
+                                  score_only=score_only)
+        if score_only and not alignment:
+            return 0
+        return alignment
+
+    def align_and_log(self, ref, query):
```

Comment:
i don't really see a benefit to doing that. it's used 4 times and makes that error handling logging pretty clean, with intermediate logging statements to indicate the direction of the alignment. 

```
LOGGER.error('Forward alignments')
self.align_and_log(ref_5p, _5p)
self.align_and_log(ref_3p, _3p)
LOGGER.error('Reverse alignments')
self.align_and_log(ref_5p_RC, _3p)
self.align_and_log(ref_3p_RC, _5p)
```

would become 

```
LOGGER.error('Forward alignments')
for pair in ((ref_5p, _5p), (ref_3p, _3p)):
        alignments = self.align(ref, query, False)
    for alignment_tuple in alignments:
        a1, a2, score, begin, end = alignment_tuple
        if self.is_high_scoring(score, query):
            LOGGER.error(format_alignment(*alignment_tuple))
LOGGER.error('Reverse alignments')
for pair in ((ref_5p_RC, _3p), (ref_3p_RC, _5p)):
        alignments = self.align(ref, query, False)
    for alignment_tuple in alignments:
        a1, a2, score, begin, end = alignment_tuple
        if self.is_high_scoring(score, query):
            LOGGER.error(format_alignment(*alignment_tuple))
```

I've also used this function elsewhere for debugging and found it helpful to have around.